src/handles/handle.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+// /// An enum repsenting different types of
+// /// handles.
+// #[derive(Debug, Copy, Clone, Eq, PartialEq)]
+// pub enum Handle {
+//     Pcr(PcrHandle),                   // Handle for pcrs.
+//     Key(KeyHandle),                   // Handle for keys.
+//     NvIndex(NvIndexHandle),           // Handle for nv indexes.
+//     TpmConstants(TpmConstantsHandle), // Handle that comes from tpm rh constants.
+//     Auth(AuthHandle),                 // Handle used for authentizations.
+//     Password,                         // The specific password handle.
+//     Object(ObjectHandle),             // General handle.
+//     None,                             // The specific None handle.
+// }
+
+// impl From<Handle> for ESYS_TR {
+//     fn from(handle: Handle) -> ESYS_TR {
+//         match handle {
+//             Handle::Pcr(handle) => handle.into(),
+//             Handle::Key(handle) => handle.into(),
+//             Handle::NvIndex(handle) => handle.into(),
+//             Handle::TpmConstants(handle) => handle.into(),
+//             Handle::Auth(handle) => handle.into(),
+//             Handle::Password => ESYS_TR_PASSWORD,
+//             Handle::Object(handle) => handle.into(),
+//             Handle::None => ESYS_TR_NONE,
+//         }
+//     }
+// }
+
+/// Macro for generating a basic handle implementation
+macro_rules! impl_basic_handle {
+    (
+        $(#[$outer:meta])*
+        $handle_type:ident
+    ) => {
+        use crate::tss2_esys::ESYS_TR;
+        use std::convert::From;
+
+        $(#[$outer])*
+        #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+        pub struct $handle_type {
+            value: u32,
+        }
+
+        impl $handle_type {
+            pub fn value(&self) -> u32 {
+                self.value
+            }
+        }
+
+        impl From<ESYS_TR> for $handle_type {
+            fn from(tss_esys_object_handle: ESYS_TR) -> $handle_type {
+                $handle_type {
+                    value: tss_esys_object_handle,
+                }
+            }
+        }
+
+        impl From<$handle_type> for ESYS_TR {
+            fn from(handle: $handle_type) -> ESYS_TR {
+                handle.value
+            }
+        }
+    };
+}
+
+/// Object handle module
+pub mod object {
+    impl_basic_handle!(
+        /// General handle type.
+        ObjectHandle
+    );
+
+    impl ObjectHandle {
+        pub const fn new(value: u32) -> ObjectHandle {
+            ObjectHandle { value }
+        }
```

Comment:
The same check should probably be done in any conversion from `ESYS_TR` too

Reply:
Actually, now that I think of it, this shouldn't really be public. `ESYS_TR` handles are only created by the ESAPI layer - we have `Copy` and `Clone` implemented on the `ObjectHandle` type, so it can be duplicated if necessary. Are there use cases where the client is the one specifying what handle an object should have?

Reply:
Your trail of thought are exactly like mine when I wrote this!

I decided to leave it there in the draft but I do not have anything against removing it.

Reply:
I don't think we need to remove it, but `pub(crate)` is probably safer - at least that way we can still use it.